src/opencl/pfx_kernel.cl

```diff
@@ -0,0 +1,168 @@
+#include "opencl_device_info.h"
+#include "opencl_misc.h"
+#include "opencl_sha1_ctx.h"
+
+#ifndef PLAINTEXT_LENGTH
+#error PLAINTEXT_LENGTH must be defined
+#endif
+
+// input
+typedef struct {
+	uint32_t length;
+	uint8_t v[PLAINTEXT_LENGTH];
+} pfx_password;
+
+// output
+typedef struct {
+	uint8_t v[20];
+} pfx_hash;
+
+// input
+typedef struct {
+	uint32_t iterations;
+	uint32_t keylen;
+	uint32_t saltlen;
+	uint8_t salt[20];
+} pfx_salt;
+
+#define PKCS12_MAX_PWDLEN 128
+
+void pkcs12_fill_buffer(unsigned char *data, size_t data_len, const unsigned char *filler, size_t fill_len)
+{
+	unsigned char *p = data;
+	size_t use_len;
+	int i;
+
+	while (data_len > 0) {
+		use_len = (data_len > fill_len) ? fill_len : data_len;
+		// memcpy(p, filler, use_len);
+		for (i = 0; i < use_len; i++)
+			p[i] = filler[i];
+		p += use_len;
+		data_len -= use_len;
+	}
+}
+
+int pkcs12_pbe_derive_key(int iterations, int id, const unsigned
+		char *pwd, size_t pwdlen, const unsigned char *salt,
+		size_t saltlen, unsigned char *key, size_t keylen)
+{
+	size_t i;
+	unsigned char unipwd[PKCS12_MAX_PWDLEN * 2 + 2], *cp = unipwd;
+
+	for (i = 0; i < pwdlen; i++) {
+		*cp++ = 0;
+		*cp++ = pwd[i];
+	}
```

Comment:
The password formatting looks similar to UTF-16 conversion but it is not the same I think. See https://tools.ietf.org/html/rfc7292#appendix-B.1 for the exact details.

I can do the conversion on CPU in `set_key()` but that would double the size of data transfers from CPU to GPU. It might not be worth it. Maybe this can be done in a different commit?